dhis2/app-platform
# feat(pwa): track online status [LIBS-315]

Adds the service worker, offline interface, and app adapter foundation for tracking the DHIS2 connection status alongside incidental requests.

Part of [LIBS-315](https://dhis2.atlassian.net/browse/LIBS-315)

### The spec
* We want to detect connection to the DHIS2 server.
* We want to minimize the network and server resources used to do so
* The service worker watches app network traffic and uses the success or failure of responses to determine connectivity status, then send messages to the client to broadcast the updates
    * Only requests to the DHIS2 server (based on the Base URL) are used to determine the connection status
* If there is no app network traffic to the DHIS2 server, the client will send pings
    * The service worker is configured to only handle pings over the network, i.e. not to cache ping requests/responses, so the result observed on the client will reflect the server connection status
    * If the connection status detected by the ping is the same as the previous status, then the time to the next ping will increase. This will continue to create an exponential back-off, which will max at out at a value, for example 5 minutes.
    * If the connection status has changed (either by ping or by app network traffic), the exponential back-off will be reset and the timeout to the next ping will be reduced back to the initial timeout.
    * If the app loses focus, a “standby” procedure is followed to stop pinging when it’s unnecessary
        * Nothing happens initially. If the user refocuses the app before the next ping is scheduled to be sent, the behavior will not change.
        * If the next-scheduled ping would take place while the app is defocused, instead the hook enters a “standby” mode.
        * When the app is refocused, if the hook is in “standby”, a ping will be sent immediately to detect the current connection status, following other rules for back-off
        * If the app is opened while defocused initially, it will also start running in standby and will not ping until refocused.
    * If the app detects an ‘offline’ event from the browser, it will ping immediately to check the connection status, since that should be a low-cost action
    * If the app detects an ‘offline’ event from the browser and is defocused, a special thing happens: it enters standby to send a ping imediately when the app is refocused again
        * If the app is refocused before the next regularly-scheduled ping, the ping timeout will be restarted but not incremented.
        * Otherwise, if the regularly scheduled ping is ready before the app is refocused, the normal standby procedure is followed (e.g. the next timeout can be incremented if the connection status hasn’t changed)

### Notes for this PR
* The service worker is extended to track requests to the DHIS2 base URL and send updates to the clients based on the successes or failures of those requests
    * This is accomplished by a plugin that’s added to all the Workbox route handlers
    * Note that the SW handles ‘ping’ requests differently: it uses only the network to avoid caching, and doesn’t use the connection status plugin, since it assumes the ping is being used by the client to determine connection status (in which case the message broadcast from the SW would be redundant)
* To avoid sending a ton of messages from the SW to every client (one for every request to the DHIS2 base url it handles), the message sending function is throttled to send max 1 message a second. Messages should still be relatively frequent though in order to inform the client to delay unnecessary pings.
* Base URLs are now stored in IndexedDB on a per-app basis
    * IndexedDB is used because it is asynchronous, so the service worker can read the base URL and detect if a request is headed for the DHIS2 server or not to determine connectivity
    * Base URLs are saved on a per-app basis because it’s technically possible to configure an  app to target a DHIS2 isntance that it is not hosted on as its base URL, and after speaking with Austin, we want to support that case
    * Much of this handling happens in the App Adapter, in the ServerVersionProvider and LoginModal components
    * The @dhis2/pwa package exports some utils for performing the IndexedDB operations for base URLs

### To test locally:
_update: some of these tools are not available any more, since much of the testing has passed_
* Check out the app-runtime branch and build it
* In the PWA app’s RequestTester.js file, follow the comments to use the useDhis2ConnectionStatus hook instead of the dummy placeholder. Note that this will be imported from app-runtime when it’s released
* Run `yarn build:pwa && yarn build:adapter`
* then run `yarn temp:start` to copy the built app-runtime libs into the app-platform node_modules and start the PWA app
* The console logs and network tab can give you some cues about what’s going on under the hood

### Todos:
- [x] remove console logs after testing
- [x] remove temp scripts

[LIBS-315]: https://dhis2.atlassian.net/browse/LIBS-315?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
